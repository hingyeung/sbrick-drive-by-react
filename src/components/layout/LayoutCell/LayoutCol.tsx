import * as React from 'react';
import * as classNames from 'classnames';
import { StatelessComponent } from 'react';

interface LayoutCellComponentProps {
  className?: string;
  componentClass?: any;
  sm?: number;
  md?: number;
  lg?: number;
  xl?: number;
}

const isValidBootstrapColumnValue = function(value: number) {
  return value >= 1 && value <= 12;
};

const buildBootstrapGridColumnClasses = function(props: any): any {
  return {
    [`col-sm-${props.sm}`]: props.sm && isValidBootstrapColumnValue(props.sm),
    [`col-md-${props.md}`]: props.md && isValidBootstrapColumnValue(props.md),
    [`col-lg-${props.lg}`]: props.lg && isValidBootstrapColumnValue(props.lg),
    [`col-xl-${props.xl}`]: props.xl && isValidBootstrapColumnValue(props.xl)
  };
};

const LayoutCol: StatelessComponent<LayoutCellComponentProps> = (props) => {
  const {
    componentClass: Component,
    className,
    ...otherProps
  } = props;

  return (
    <Component {...otherProps} className={classNames(className, 'col', buildBootstrapGridColumnClasses(otherProps))}/>
  );
};

LayoutCol.defaultProps = {
  componentClass: 'div'
};

export default LayoutCol;