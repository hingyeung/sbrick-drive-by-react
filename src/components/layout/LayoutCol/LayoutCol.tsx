import * as React from 'react';
import * as classNames from 'classnames';

interface LayoutColComponentProps {
  // className?: string;
  componentClass?: any;
  sm?: number;
  md?: number;
  lg?: number;
  xl?: number;
  propsForComponent?: object;
}

const isValidBootstrapColumnValue = function (value: number) {
  return value >= 1 && value <= 12;
};

const buildBootstrapGridColumnClasses = function (props: any): any {
  return {
    [`col-sm-${props.sm}`]: props.sm && isValidBootstrapColumnValue(props.sm),
    [`col-md-${props.md}`]: props.md && isValidBootstrapColumnValue(props.md),
    [`col-lg-${props.lg}`]: props.lg && isValidBootstrapColumnValue(props.lg),
    [`col-xl-${props.xl}`]: props.xl && isValidBootstrapColumnValue(props.xl)
  };
};

const LayoutCol: React.SFC<LayoutColComponentProps &
  React.HTMLAttributes<HTMLDivElement> &
  React.ClassAttributes<HTMLDivElement>> = (props) => {
  const {
    componentClass: Component,
    className,
    propsForComponent,
    ...otherProps,
  } = props;

  return (
    <Component
      {...propsForComponent}
      className={classNames(className, 'col', buildBootstrapGridColumnClasses(otherProps))}
      {...otherProps}
    />
  );
};

LayoutCol.defaultProps = {
  componentClass: 'div'
};

export default LayoutCol;