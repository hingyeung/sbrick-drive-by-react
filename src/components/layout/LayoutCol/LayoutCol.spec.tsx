import * as React from 'react';
import LayoutCol from './LayoutCol';
import { shallow } from 'enzyme';

describe('LayoutColComponent', () => {
  const hasClassWithPrefix = (wrapper: any, prefix: string) => {
    expect(
      wrapper.find('.hello').prop('className').trim().split(/\s+/).filter((thisClass: string) =>
        thisClass.indexOf(prefix) > -1)
    ).toEqual([]);
  };
  const sizes = ['sm', 'md', 'lg', 'xl'];

  it('should add "row" class to specified component', () => {
    const wrapper = shallow(<LayoutCol className="extra-class">hello</LayoutCol>);
    expect(wrapper.find('.col')).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('should use componentClass in props as children', () => {
    const wrapper = shallow(<LayoutCol className="extra-class" componentClass={'b'}/>);
    expect(wrapper.find('b').exists()).toBeTruthy();
    expect(wrapper.find('b.col.extra-class')).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('should pass other props to componentClass', () => {
    const wrapper = shallow(
      <LayoutCol
        className="extra-class"
        componentClass={'b'}
        propsForComponent={{A: 'a', B: 'b'}}
      />);
    expect(wrapper.find('b').prop('A')).toEqual('a');
    expect(wrapper.find('b').prop('B')).toEqual('b');
    expect(wrapper).toMatchSnapshot();
  });

  sizes.forEach(size =>
    [1, 12].forEach(width =>
      it(`should set ${size} screen column number when ${size} prop is set`, () => {
        const sizeProps = {[size]: width};
        const wrapper = shallow(
          <LayoutCol className="hello" {...sizeProps}>Hello</LayoutCol>
        );

        expect(wrapper.find('.hello').hasClass(`col-${size}-${width}`)).toBeTruthy();
      })
    )
  );

  sizes.forEach(size =>
    [0, 13].forEach(width =>
      it('should not set ${size} screen column number when ${size} prop out of range', () => {
        const sizeProps = {[size]: width};
        const wrapper = shallow(
          <LayoutCol className="hello" {...sizeProps}>Hello</LayoutCol>
        );

        expect(hasClassWithPrefix(wrapper, `col-${size}-`)).toBeFalsy();
      })
    )
  );
});