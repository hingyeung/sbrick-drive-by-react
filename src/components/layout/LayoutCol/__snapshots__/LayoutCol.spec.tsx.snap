// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutColComponent should add "row" class to specified component 1`] = `
<div
  className="extra-class col"
>
  hello
</div>
`;

exports[`LayoutColComponent should pass other props to componentClass 1`] = `
<b
  A="a"
  B="b"
  className="extra-class col"
/>
`;

exports[`LayoutColComponent should use componentClass in props as children 1`] = `
<b
  className="extra-class col"
/>
`;
